### ISSUE TITLE
Support for iterators within HDF5::Files or Groups

### ISSUE THREAD

**Issue by MPhysXDev** [dealii/dealii#9043]
Dear deal.II developers,

the initial support for HDF5 in v9.1 is a great addition, specially in the context of python script interactions with deal.II. 

However adding interfaces to the built-in iterators over hdf5 groups or attributes (see for instance the h5py python interface package for comparison) would be a VAST improvement opening up greatly the use cases...

If this is already in a development branch, or contemplated within a time frame you feel comfortable to share, we would be extremely interested to know.

Best regards,

MPhysX Development Team

**Comment by bangerth** [comment]
I think this would be a question for @dangars -- though I suspect that the answer is that no iterators are currently planned. Obviously, however, any help with writing such iterators would be greatly appreciated!

**Comment by dangars** [comment]
I tried to keep the HDF5 interface as close as possible to hd5py which is a great library!
I agree that iterators would be a vast improvement.
I'm going to write the feature that you requested, iterators to be able to loop over the groups and attributes. When I have a proof of concept I'll ping you!

**Comment by MPhysXDev** [comment]
@dangars, great ! We are looking forward to hear back about your progress on this. h5py is really a good one indeed.

Best regards,

MPhysX Development Team

**Comment by dangars** [comment]
@MPhysXDev I gave it a another thought, and I realized that it is not possible to add iterators in the same way that h5py works.

A HDF5 attribute, can be a double, int, string, a matrix, compound type, etc. And there is no way to know at compile time the type of a particular HDF5 attribute.

Python is a dynamic typed language, this allows to create the iterators that you find in the h5py library. You can loop over the attributes and h5py can return an int, double, string or something else. Then at run time python can figure out how to deal with the particular type of the attribute.

In the other hand C++ is a static typed language. The return type of a function has to be known at compile time. As you can see below I have to explicitly tell the HDF5 interface what is the type of the attribute (in this case a double)
```c++
get_attribute<double>("attribute")
```
If `attribute` is not a double in the HDF5 file, then the HDF5 library can try to do a type conversion. For example `HDF5` can convert from `int` to `double`, but if `attribute` is a matrix then `HDF5` will raise an error.

Should we close this issue, as I don't think that it is feasible to recreate the python `h5py` functionality in C++.

**Comment by bangerth** [comment]
You can let the iterator return a `boost::any` object that you initialize based on the type of the attribute. (I assume you can query HDF5 for the type, or at least ask whether it is of a specific type.) The user then has to know what type a particular attribute has to decode the `boost::any`, but at least you can write iterators.

**Comment by dangars** [comment]
> You can let the iterator return a `boost::any` object 

@bangerth That is a good point :)
boost::any or std::any (C++17) can be used in this context. The HDF5 library can be used to query the type. Then `std::any::type()` can be used to query the contained type.

Although I think that I would prefer `std::variant` (C++17) it seems safer to me as it explicitly states the list of allowed types and it is easier to use it in a templated environment. `std::holds_alternative<type>(v)` is used to query the contained type.

**Comment by masterleinad** [comment]
Make sure that it works both for `boost::variant` and `std::variant` such that we can use a compatibility wrapper.

**Comment by dangars** [comment]
> Make sure that it works both for `boost::variant` and `std::variant` such that we can use a compatibility wrapper.

Sounds good!

**Comment by bangerth** [comment]
Yes, if you can enumerate the possible types of an attribute, then `***::variant` is the right choice!

If you want to write things in a generic way, take a look at what we have done in `include/deal.II/base/std_cxx17/optional.h`. It should be easy to replicate this for `std::variant`.
